src/lib/Libdb/pgsql/pbs_db_utility

```diff
@@ -0,0 +1,1302 @@
+		old_inst_dir="${PBS_HOME}/pgsql.old"
+		old_data_dir="${PBS_HOME}/datastore.old"
+	elif [ -d "${PBS_HOME}/pgsql.forupgrade" ]; then
+		old_inst_dir="${PBS_HOME}/pgsql.forupgrade"
+		old_data_dir="${PBS_HOME}/datastore.forupgrade"
+	else
+		echo "Data service directory from previous PBS installation not found,"
+		echo "Datastore upgrade cannot continue"
+		return 1
+	fi
+
+	# strip the minor version from sys_pgsql_ver if old_pgsql_ver does not have minor version (for comparison).
+	[[ ! $old_pgsql_ver =~ "." ]] && sys_pgsql_ver=$(echo $sys_pgsql_ver | cut -d '.' -f 1)
+
+	[ ${sys_pgsql_ver%.*} -eq ${old_pgsql_ver%.*} ] && [ ${sys_pgsql_ver#*.} \> ${old_pgsql_ver#*.} ] || [ ${sys_pgsql_ver%.*} -gt ${old_pgsql_ver%.*} ];
+	result=$?
+	if [ ${result} -eq 0 ]; then
+		if [ -d "$PBS_EXEC/pgsql" ]; then
+			#Start upgrade process of datastore
+			upgrade_db
+			return $?
+		else
+			return 2
+		fi
+	elif [ "${old_pgsql_ver}" = "${sys_pgsql_ver}" ]; then
+		return 0
+	fi
+
+	echo "Upgrade from version ${old_pgsql_ver} unsupported"
+	return 1
+}
+
+upgrade_db() {
+	#
+	# This routine will insatll a 9.3 database cluster,
+	# will perform the pre-upgrade checks for datastore
+	# with appropriate authentication management.
+	#
+
+	server_ctl="${PBS_EXEC}/sbin/pbs_dataservice"
+	if [ ! -x "${server_ctl}" ]; then
+		echo "${server_ctl} not found"
+		return 1
+	fi
+
+	if [ ! -x "${PBS_EXEC}/sbin/pbs_ds_password" ]; then
+		echo "${PBS_EXEC}/sbin/pbs_ds_password not found"
+		return 1
+	fi
+
+	if [ ! -x "${inst_dir}/bin/pg_upgrade" ]; then
+		echo "${inst_dir}/bin/pg_upgrade not found"
+		return 1
+	fi
+
+	# Check and convert UTF-8 character encoded database to SQL_ASCII
+	utf8_to_sql_ascii
+	if [ $? -ne 0 ]; then
+		return 1
+	fi
+
+	# Backup datastore directory, if backup directory already
+	# present then exit.
+	if [ -d "${old_data_dir}" ]; then
+		echo "Files from previous datastore upgrade found,"
+		echo "Datastore upgrade cannot continue"
+		return 1
+	else
+		err=`mv ${data_dir} ${old_data_dir} 2>&1`
+		if [ $? -ne 0 ]; then
+			echo "${err}"
+			return 1
+		fi
+	fi
+
+	# Invoke the dataservice creation script for pbs
+
+	upgrade=1
+	pbs_install_db
+	ret=$?
+
+	if [ ${ret} -ne 0 ]; then
+		echo "*** Error initializing the PBS dataservice"
+		echo "Error details:"
+		echo "$resp"
+		return ${ret}
+	fi
+
+	# Copy the pg_hba.conf from old cluster.
+	err=`cp -p ${old_data_dir}/pg_hba.conf ${data_dir}/pg_hba.conf`
+	if [ $? -ne 0 ]; then
+		echo "${err}"
+		return 1
+	fi
+
+	set_db_trust_login "${data_dir}"
+	ret=$?
+	if [ $ret -ne 0 ]; then
+		return $ret
+	fi
+	set_db_trust_login "${old_data_dir}"
+	ret=$?
+	if [ $ret -ne 0 ]; then
+		return $ret
+	fi
+
+	CWD=`pwd`
+	cd "${data_dir}"
+	#Perform pg_upgrade -c to check if we can upgrade the cluster or not
+	err=`su ${user} -c "/bin/sh -c '${PGSQL_LIBSTR} ${inst_dir}/bin/pg_upgrade -b ${old_inst_dir}/bin -B ${inst_dir}/bin -d ${old_data_dir} -D ${data_dir} -c'" 2>&1`
```

Comment:
When we move to PG 10 in the next few months, pg_upgrade will not work, and we will have to do a dump and restore upgrade to support overlay upgrades. Is this current code structure good for such a upgrade?

Reply:
We do a pg_upgrade -c, which does the simulation of upgrade and if it fails maybe we can proceed with a dump and restore upgrade.